### DISCUSSION TITLE
Webui crashed today

### DISCUSSION THREAD

**Discussion by ahroyce** [Gerschel/sd_web_ui_preset_utils#29] [General]
Tried starting Webui today and after it went through a git pull it does this. and if I press any key it crashes...Thoughts?
Already up to date.
venv "C:\Users\aroyc\stable-diffusion-webui\venv\Scripts\Python.exe"
Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep 5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)]
Commit hash: https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/ff6a5bcec1ce25aa8f08b157ea957d764be23d8d
Installing requirements for Web UI
Launching Web UI with arguments: --xformers
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
Loading weights [5d5ad06cc2] from C:\Users\aroyc\stable-diffusion-webui\models\Stable-diffusion\mdjrny-v4.ckpt
Applying xformers cross attention optimization.
Textual inversion embeddings loaded(0):
Textual inversion embeddings skipped(7): anthro, CGI_Animation-185, CGI_Animation-245, CGI_Animation, midjourney, VikingPunk, vray-render
Model loaded in 5.9s (0.6s create model, 3.2s load weights).
Traceback (most recent call last):
File "C:\Users\aroyc\stable-diffusion-webui\launch.py", line 308, in
start()
File "C:\Users\aroyc\stable-diffusion-webui\launch.py", line 303, in start
webui.webui()
File "C:\Users\aroyc\stable-diffusion-webui\webui.py", line 159, in webui
shared.demo = modules.ui.create_ui()
File "C:\Users\aroyc\stable-diffusion-webui\modules\ui.py", line 588, in create_ui
modules.scripts.scripts_txt2img.initialize_scripts(is_img2img=False)
File "C:\Users\aroyc\stable-diffusion-webui\modules\scripts.py", line 245, in initialize_scripts
script = script_class()
File "C:\Users\aroyc\stable-diffusion-webui\extensions\sd_web_ui_preset_utils\scripts\zpreset_utils.py", line 119, in init
self.positon_manager = self.reorder_map[self.ui_first]
KeyError: 'inpaint'
Press any key to continue . . .

I moved your extension folder to solve this problem.  Looks like the latest update knocked your code out.

 I really like what your Preset tool does I found it to work very nicely.  I do have a question however, I created several presets and noted that when I created a certain amount of them that the window that displayed the list of presets did not show a scroll bar when the number of presets I created exceeded the size of the preset window.

**Comment by Gerschel** [comment]
I pushed an update. You might have to go in to the extension and do a git pull yourself. You could also delete the file zpreset_utils.py, and do the update process in the extensions tab.  

They modified how the sorter works, I deleted eveything with that process, so it shouldn't happen again.  

Navigate to this folder, you can copy and paste this path, it's based on your setup, and delete zpreset_utils.py
```bash
C:\Users\aroyc\stable-diffusion-webui\extensions\sd_web_ui_preset_utils\scripts\
```
Or you could even delete lines 117, 118, 119, and it will run.

**Reply by ahroyce** [reply]
That worked...thank you for timely reply

**Comment by Gerschel** [comment]
I have some questions. I've done a full rewrite, it's not polished. I made it able to work on all tabs. In order for me to see it on all tabs, I pulled it out of the flow and turned it into a floating window. But, to me it feels like it's in the way because everything flows to fill the void.  
So I made it collapse, and it drops it's opacity when not hovered over.  

![image](https://user-images.githubusercontent.com/9631031/212831675-bdfa887f-a605-4b58-8b34-3b47bf7256eb.png)


![image](https://user-images.githubusercontent.com/9631031/212831759-6eaab12a-ac92-4511-a3cb-d5a9a4d88e0f.png)


And, now it reaches all components, except a few that share the same element id in the same region (region being quicksetting, scripts, and tabs main area)  

![image](https://user-images.githubusercontent.com/9631031/212832064-52c2b6f4-cda2-4e13-9e14-ee521d6f013c.png)  


But I lost the ability for a preset to be used from between txt2img and img2img.  

The question is, if a preset was saved, and it included quicksettings, should the quicksettings be separate so it can be used across tabs, or is it no big deal because it would be used for those specific settings anyways?  

This dilemma is keeping me from finishing it. I don't like how the checkbox group is so big, then when you go on deforum tab, it's huge.  

![image](https://user-images.githubusercontent.com/9631031/212832780-3da565dd-2a81-4af0-bf88-c9aaf5c6594d.png)  

Which is what makes me consider having more separation by regions and splitting up the checkbox group. It's hard to do want to do a partial rewrite when feel like I have so much time on it, I just need to get to a release version. But if I release it too early, then I'll probably get a lot of support requests while I'm trying to handle the improvement. Right now I have a bit of time, I'm a Carpenter going through my layoff period, but soon I will be so busy I wont get a day off for months.  

Doing the separation will change my data structure, and I can do a reasonable hack in a fair amount of time, but would like to not uglify my code too soon, so would rather have a clean structure and do it right.  

If you can give me some insight to how an expectation of workflow would look like, that would help.

**Reply by hzielke** [reply]
Interesting idea, using a floating window. Regarding your question: Personally, I did not use quicksettings across tabs yet, so no big deal for me. Regarding the checkbox group - would it be either possible to do some nesting expansions like on the settings page or perhaps a settings page itself to choose which checkboxes to display? I think I'd like nestings better for my workflow.

**Reply by ahroyce** [reply]
I like the idea of a clean structure and I like what Hzielke suggested nesting expansions.

**Reply by Gerschel** [reply]
I'm with this as well. So far I have three different regions, the quicksettings, tab main area, and scripts if scripts are available on tab. I've made a class that can serve for each region, but can also have sub-regions. At his stage, I would have to hardcode what constitutes as a sub-region. The first iteration of breaking up into sub-regions, would be tabs that belong to a tab. If I get this right, I may be able to have further breaking up. Thanks for the feedback.

**Reply by inewlife** [reply]
Hi, Could reply those code?

**Reply by Gerschel** [reply]
@inewlife 
Were you asking about the version 2 I had for a bit?
If so, it'll need a major overhaul.  

The way it worked, it read from gradio app configs, seeked out each html element that the config was aware of, and categorized them, then it had handlers that operated with the different types.  

There were some changes to gradio's configs, that are no big deal to update to.  
The main issue became, as the webui was getting built up, people implemented elements in various ways. 

This meant I would have to code for every way someone wrote an element, but they also used the wrong convention to state what the elements were, they would declare it was a dropdown when it was an entry, button when it's a checkbox, and so forth.  

While I am capable of doing a full overhaul, every time a script kiddie writes in a non-standard implementation, it would break things.  

The only ways to get around it: 
code it to release versions only, 
flat out reject support on bad code and ask the community to put pressure on them to fix their stuff,
or put together my own release versions that actually fix the issues and give more control to the user.  


One of them is a lot of work with no pay, and when in real life work demands my attention, I can't afford the time to do it.  
Two of them will have a lot of users that don't understand the situation and think the issue is in how I implemented my code.  

If someone was willing to compensate me for my time to build any of these out, then this can become a real job, and I am capable of handling it.  

I still have the core code that still works to some degree, and I've already determined what I needed to do to have it updated. I just haven't updated it because it didn't seem to be worth the effort, that inevitably, they'll implement full control of the entire app with presets.

**Reply by inewlife** [reply]
Can you tell me about how to switch the chekpoint model?
